expirationd.lua

```diff
@@ -215,14 +206,18 @@ local Task_methods = {
     stop = function (self)
         if (get_fiber_id(self.guardian_fiber) ~= 0) then
             self.guardian_fiber:cancel()
-            while self.guardian_fiber:status() ~= 'dead' do
+            while self.guardian_fiber:status() ~= "dead" do
                 fiber.sleep(0.01)
             end
             self.guardian_fiber = nil
         end
         if (get_fiber_id(self.worker_fiber) ~= 0) then
-            self.worker_fiber:cancel()
-            while self.worker_fiber:status() ~= 'dead' do
+            if self.atomic_iteration then
+                self.worker_canceled = true
+            else
+                self.worker_fiber:cancel()
+            end
+            while self.worker_fiber:status() ~= "dead" do
                 fiber.sleep(0.01)
             end
```

Comment:
Why not `self.worker_fiber:set_joinable(true)` + `self.worker_fiber:join()`?

Reply:
I get a weird vinyl and join bug. In some tests data is not deleted here:
```lua
-- default process_expired_tuple function
local function default_tuple_drop(space_id, args, tuple)
    box.space[space_id]:delete(construct_key(space_id, tuple))
end
```

But if we write some sleep, everything work fine:
```lua
-- default process_expired_tuple function
local function default_tuple_drop(space_id, args, tuple)
    box.space[space_id]:delete(construct_key(space_id, tuple))
    fiber.sleep(0.1)
end
```

Reply:
How is it related to my comment?

```
while self.worker_fiber:status() ~= "dead" do
                fiber.sleep(0.01)
            end
```

Here you wait until fiber is terminated. I suggest to do it without a loop and use fiber.join.

Reply:
Yes, I say that I changed loop to fiber.join and got this bug https://github.com/tarantool/expirationd/pull/52#discussion_r662213292

Reply:
It's quite strange. Could it be workarounded for at test level?

Reply:
Insert custom deleter, but this is wierd, can we just leave loop with sleep?

Reply:
Anyway you put such code without any comment and how should people in future understand it?

Reply:
I have helpers.retry to check the result, sleep does not work in the test, everything breaks in the section with deletion. What code have I put without comments and that it is not clear? sleep with a status check for waiting for a fiber was in the original implementation. Perhaps somewhere additional comments are needed, but I would like to know which place you specifically point to

Reply:
Before I thought that this code really simply wait when fiber will be finished. But as I understand you also state that this code makes vinyl tests work. It's strange, not obvious.

Reply:
Also with
```patch
diff --git a/expirationd.lua b/expirationd.lua
index 34367b6..0938702 100644
--- a/expirationd.lua
+++ b/expirationd.lua
@@ -205,21 +205,19 @@ local Task_methods = {
     end,
     stop = function (self)
         if (get_fiber_id(self.guardian_fiber) ~= 0) then
+            self.guardian_fiber:set_joinable(true)
             self.guardian_fiber:cancel()
-            while self.guardian_fiber:status() ~= "dead" do
-                fiber.sleep(0.01)
-            end
+            self.guardian_fiber:join()
             self.guardian_fiber = nil
         end
         if (get_fiber_id(self.worker_fiber) ~= 0) then
+            self.worker_fiber:set_joinable(true)
             if self.atomic_iteration then
                 self.worker_canceled = true
             else
                 self.worker_fiber:cancel()
             end
-            while self.worker_fiber:status() ~= "dead" do
-                fiber.sleep(0.01)
-            end
+            self.worker_fiber:join()
             self.worker_fiber = nil
         end
     end,
```

I have:
```bash
➜  expirationd git:(indexpiration) ✗ .rocks/bin/luatest -v
Started on Thu Jul  1 16:43:19 2021
    atomic_iteration.test_passing ... (1.005s) Ok
    atomic_iteration.test_memtx ... (4.018s) Ok
    atomic_iteration.test_mvcc_vinyl_tx_conflict ... (4.071s) Ok
    atomic_iteration.test_kill_task ... (0.275s) Ok
    custom_index.test_passing ... (0.001s) Ok
    custom_index.test_tree_index ... (0.413s) Ok
    custom_index.test_tree_index_non_unique ... (0.208s) Ok
    custom_index.test_tree_index_multipart ... (0.212s) Ok
    custom_index.test_tree_index_json_path ... (0.210s) Ok
    custom_index.test_tree_index_multikey ... (2.009s) Ok
    custom_index.test_memtx_tree_functional_index ... (1.037s) Ok
    custom_index.test_hash_index ... (0.210s) Ok
    custom_index.test_hash_index_multipart ... (0.107s) Ok
    iterate_with.test_passing ... (0.002s) Ok
    iterator_type.test_passing_errors ... (0.001s) Ok
    iterator_type.test_passing_all ... (0.002s) Ok
    iterator_type.test_passing_eq ... (0.001s) Ok
    iterator_type.test_passing_gt ... (0.001s) Ok
    iterator_type.test_passing_req ... (0.001s) Ok
    iterator_type.test_passing_ge ... (0.001s) Ok
    iterator_type.test_passing_lt ... (0.001s) Ok
    iterator_type.test_passing_le ... (0.001s) Ok
    iterator_type.test_tree_index_all ... (0.413s) Ok
    iterator_type.test_hash_index_all ... (0.208s) Ok
    iterator_type.test_tree_index_eq ... (0.413s) Ok
    iterator_type.test_hash_index_eq ... (1.007s) Ok
    iterator_type.test_tree_index_gt ... (2.213s) Ok
    iterator_type.test_hash_index_gt ... (1.130s) Ok
    iterator_type.test_tree_index_req ... (0.409s) Ok
    iterator_type.test_tree_index_ge ... (0.409s) Ok
    iterator_type.test_tree_index_lt ... (2.216s) Ok
    iterator_type.test_tree_index_le ... (0.419s) Ok
    process_while.test_passing ... (0.003s) Ok
    process_while.test_tree_index ... (2.007s) Ok
    process_while.test_hash_index ... (1.006s) Ok
    start_key.test_passing ... (0.002s) Ok
    start_key.test_tree_index ... (0.622s) Ok
    start_key.test_hash_index ... (0.308s) Ok
=========================================================
Ran 38 tests in 26.612 seconds, 38 successes, 0 failures
```

It seems to be ok.
```
Tarantool Enterprise 2.8.1-0-ge2a1ec0c2-r407-macos
Target: Darwin-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/opt/actions-runners/sdk_01/_work/sdk/sdk/build.sdk/tarantool-2.8/static-build/tarantool-prefix -DENABLE_BACKTRACE=TRUE
Compiler: /Library/Developer/CommandLineTools/usr/bin/cc /Library/Developer/CommandLineTools/usr/bin/c++
C_FLAGS: -Wno-unknown-pragmas -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-gnu-alignof-expression
CXX_FLAGS: -Wno-unknown-pragmas -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-invalid-offsetof -Wno-gnu-alignof-expression
```

Reply:
I did as indicated in the documentation, maybe I have an old version of the tarantula and some bug is not fixed. But writing set_joinable like you made it work. But this doesn't work

```patch
diff --git a/expirationd.lua b/expirationd.lua
index 34367b6..bfc3f84 100644
--- a/expirationd.lua
+++ b/expirationd.lua
@@ -179,7 +179,8 @@ local function guardian_loop(task)
         -- if fiber doesn't exist
         if get_fiber_id(task.worker_fiber) == 0 then
             -- create worker fiber
-            task.worker_fiber = fiber.create(worker_loop, task)
+            task.worker_fiber = fiber.new(worker_loop, task)
+            task.worker_fiber:set_joinable(true)
 
             log.info("expiration: task %q restarted", task.name)
             task.restarts = task.restarts + 1
@@ -201,14 +202,13 @@ end
 local Task_methods = {
     start = function (self)
         self:stop()
-        self.guardian_fiber = fiber.create(guardian_loop, self)
+        self.guardian_fiber = fiber.new(guardian_loop, self)
+        self.guardian_fiber:set_joinable(true)
     end,
     stop = function (self)
         if (get_fiber_id(self.guardian_fiber) ~= 0) then
             self.guardian_fiber:cancel()
-            while self.guardian_fiber:status() ~= "dead" do
-                fiber.sleep(0.01)
-            end
+            self.guardian_fiber:join()
             self.guardian_fiber = nil
         end
         if (get_fiber_id(self.worker_fiber) ~= 0) then
@@ -217,9 +217,7 @@ local Task_methods = {
             else
                 self.worker_fiber:cancel()
             end
-            while self.worker_fiber:status() ~= "dead" do
-                fiber.sleep(0.01)
-            end
+            self.worker_fiber:join()
             self.worker_fiber = nil
         end
     end,
```
 TESTS:
```
Started on Thu Jul  1 16:51:50 2021
    atomic_iteration.test_passing ... (2.010s) Ok
    atomic_iteration.test_memtx ... (6.021s) Ok
    atomic_iteration.test_mvcc_vinyl_tx_conflict ... (8.259s) fail
...arantool/expirationd/test/unit/atomic_iteration_test.lua:153: expected:
{}
actual:
{
    [1, '1', null, null, 33],
    [2, '2', null, null, 33],
    [3, '3', null, null, 33],
    [4, '4', null, null, 33],
    [5, '5', null, null, 33],
    [6, '6', null, null, 33],
    [7, '7', null, null, 33],
    [8, '8', null, null, 32],
    [9, '9', null, null, 32],
    [10, '10', null, null, 33],
}
    atomic_iteration.test_kill_task ... (1.574s) Ok
    custom_index.test_passing ... (3.009s) Ok
    custom_index.test_tree_index ... (0.412s) Ok
    custom_index.test_tree_index_non_unique ... (0.205s) Ok
    custom_index.test_tree_index_multipart ... (0.206s) Ok
    custom_index.test_tree_index_json_path ... (0.206s) Ok
    custom_index.test_tree_index_multikey ... (5.097s) fail
...in/tarantool/expirationd/test/unit/custom_index_test.lua:172: expected:
{{1, "1", [7] = {data = {{name = "A"}, {name = "B"}}, extra_field = 1}}}
actual:
{
    [1, '1', null, null, null, null, {'data': [{'name': 'A'}, {'name': 'B'}], 'extra_field': 1}],
    [1, '1', null, null, null, null, {'data': [{'name': 'A'}, {'name': 'B'}], 'extra_field': 1}],
}
    custom_index.test_memtx_tree_functional_index ... (0.105s) Ok
    custom_index.test_hash_index ... (0.209s) Ok
    custom_index.test_hash_index_multipart ... (0.104s) Ok
    iterate_with.test_passing ... (1.007s) Ok
    iterator_type.test_passing_errors ... (0.001s) Ok
    iterator_type.test_passing_all ... (4.009s) Ok
    iterator_type.test_passing_eq ... (2.011s) Ok
    iterator_type.test_passing_gt ... (2.006s) Ok
    iterator_type.test_passing_req ... (1.002s) Ok
    iterator_type.test_passing_ge ... (1.003s) Ok
    iterator_type.test_passing_lt ... (1.006s) Ok
    iterator_type.test_passing_le ... (1.005s) Ok
    iterator_type.test_tree_index_all ... (0.410s) Ok
    iterator_type.test_hash_index_all ... (0.205s) Ok
    iterator_type.test_tree_index_eq ... (0.408s) Ok
    iterator_type.test_hash_index_eq ... (0.104s) Ok
    iterator_type.test_tree_index_gt ... (0.412s) Ok
    iterator_type.test_hash_index_gt ... (1.137s) Ok
    iterator_type.test_tree_index_req ... (0.415s) Ok
    iterator_type.test_tree_index_ge ... (0.411s) Ok
    iterator_type.test_tree_index_lt ... (0.408s) Ok
    iterator_type.test_tree_index_le ... (0.408s) Ok
    process_while.test_passing ... (2.008s) Ok
    process_while.test_tree_index ... (0.207s) Ok
    process_while.test_hash_index ... (0.103s) Ok
    start_key.test_passing ... (4.011s) Ok
    start_key.test_tree_index ... (0.621s) Ok
    start_key.test_hash_index ... (0.314s) Ok
=========================================================
```

TARANTOOL:
```
Tarantool Enterprise 2.5.3-0-gf93e48013-r393-macos
Target: Darwin-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/opt/actions-runners/sdk_01/_work/sdk/sdk/build.sdk/tarantool-2.5/static-build/tarantool-prefix -DENABLE_BACKTRACE=TRUE
Compiler: /Library/Developer/CommandLineTools/usr/bin/cc /Library/Developer/CommandLineTools/usr/bin/c++
C_FLAGS: -Wno-unknown-pragmas -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-gnu-alignof-expression
CXX_FLAGS: -Wno-unknown-pragmas -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-invalid-offsetof -Wno-gnu-alignof-expression
```

Reply:
I think the problem is in fiber.new, changing back to create everything worked for my above

Reply:
Well... Ok. Let's stay as is.